cloudinit/sources/DataSourceEc2.py

```diff
@@ -125,12 +127,30 @@ def _get_data(self):
                 LOG.debug("FreeBSD doesn't support running dhclient with -sf")
                 return False
             try:
-                with EphemeralDHCPv4(self.fallback_interface):
+                # ipv6 dualstack might succeed when dhcp4 fails
```

Comment:
This whole block feels a bit awkward, and I have a feeling we're going to be doing the same thing for other datasources in the near future. What if we have something like a `EphemeralNetwork` (or named whatever) context manager that takes arguments whether to use v4 and v6. This logic could move in there, and then the call logic could be a lot cleaner.